File: media/movies.py
"""
media/movies.py
---------------
Movie-specific parsing, serialization, and DB logic for MOASYS-Vault.

Expected Plex folder structure:
  <quality_folder>/
    <Movie Title (YEAR)>/
      <Movie Title (YEAR)>.mp4
      <Movie Title (YEAR)> {edition-Edition Name}.mp4
"""

import os
import re
import json
from datetime import datetime

# ─────────────────────────────────────────────
# Constants
# ─────────────────────────────────────────────

EARLIEST_FILM_YEAR = 1888                  # Roundhay Garden Scene — earliest known film
CURRENT_YEAR       = datetime.now().year   # Evaluated once at startup

# ─────────────────────────────────────────────
# Regex patterns
# ─────────────────────────────────────────────

# Matches Plex-style movie folder names: "The Crow (1994)"
# Group 1 = title, Group 2 = year
FOLDER_PATTERN = re.compile(r'^(.+)\s\((\d{4})\)$')

# Matches Plex-style movie file stems, with an optional edition tag:
#   "The Crow (1994)"
#   "The Crow (1994) {edition-Director's Cut}"
#   "The Crow (1994) {edition-}"   ← empty edition, caught as a warning
# Group 1 = title, Group 2 = year, Group 3 = edition (or None if no tag)
# Note: [^}]* allows zero characters so empty {edition-} is captured rather than rejected
FILE_PATTERN = re.compile(r'^(.+)\s\((\d{4})\)(?:\s\{edition-([^}]*)\})?$')


# ─────────────────────────────────────────────
# Helpers
# ─────────────────────────────────────────────

def _is_video(filename, media_config):
    """Return True if the file extension is in the configured video_extensions list."""
    _, ext = os.path.splitext(filename)
    return ext.lower() in [e.lower() for e in media_config["video_extensions"]]

def _get_primary_extensions(media_config):
    """
    Return the primary extension(s) as a lowercase list.
    config value must be a list e.g. [".mp4"] or [".mp4", ".mkv"].
    """
    return [e.lower() for e in media_config["primary_extension"]]

def _is_primary(filename, media_config):
    """Return True if the file extension matches any configured primary_extension."""
    _, ext = os.path.splitext(filename)
    return ext.lower() in _get_primary_extensions(media_config)

def _format_primary_exts(media_config):
    """
    Return a human-readable string of primary extensions for warning messages.
    e.g. ".mp4" -> "Non-.MP4"
         [".mp4", ".mkv"] -> "Non-.MP4/.MKV"
    """
    exts = _get_primary_extensions(media_config)
    return "Non-" + "/".join(e.upper() for e in exts)

def _parse_folder(name):
    """
    Parse a Plex movie folder name into (title, year).
    Returns None if the name doesn't match the expected format.
    Example: "The Crow (1994)" -> ("The Crow", 1994)
    """
    m = FOLDER_PATTERN.match(name)
    return (m.group(1).strip(), int(m.group(2))) if m else None

def _parse_file_stem(stem):
    """
    Parse a Plex movie file stem (filename without extension) into (title, year, edition).
    Returns None if the stem doesn't match the expected format.

    edition is:
      None  — no {edition-...} tag present
      ""    — tag present but empty: {edition-}
      str   — normal edition name e.g. "Director's Cut"
    """
    m = FILE_PATTERN.match(stem)
    if m:
        edition_raw = m.group(3)
        if edition_raw is None:
            edition = None             # No edition tag at all
        elif edition_raw.strip() == "":
            edition = ""               # Empty edition tag: {edition-}
        else:
            edition = edition_raw.strip()
        return (m.group(1).strip(), int(m.group(2)), edition)
    return None

def _make_key(title, year, edition):
    """
    Build a unique dictionary key for a movie record.
    Lowercased so "The Crow" and "the crow" are treated as the same title.
    This key is used to merge duplicate entries across quality folders.
    """
    return (title.lower(), year, (edition or "").lower())


# ─────────────────────────────────────────────
# Scanner (called by core/scanner.py)
# ─────────────────────────────────────────────

def scan_quality_folder(quality_path, folder_name, tag, media_config, warnings):
    """
    Walk one quality folder (e.g. UHD/) and return a dict of movie records.
    Called once per quality folder by core/scanner.py.

    Returns: { key: { title, year, edition, qualities: set() } }
    The qualities set starts with just this folder's tag (e.g. {"UHD"}).
    core/scanner.py merges records from multiple quality folders together.
    """
    records = {}

    # Each subfolder inside a quality folder should be a movie folder
    for movie_folder in os.scandir(quality_path):
        if not movie_folder.is_dir():
            continue   # Skip any loose files sitting in the quality folder

        folder_rel    = os.path.join(folder_name, movie_folder.name)   # e.g. "UHD/The Crow (1994)"
        parsed_folder = _parse_folder(movie_folder.name)

        # Read all files inside this movie folder
        try:
            all_files = list(os.scandir(movie_folder.path))
        except PermissionError:
            warnings.add(folder_rel, "Permission denied reading folder")
            continue

        # Separate files into video and non-video, then primary and non-primary
        video_files   = [f for f in all_files if _is_video(f.name, media_config)]
        non_primary   = [f for f in video_files if not _is_primary(f.name, media_config)]
        primary_files = [f for f in video_files if _is_primary(f.name, media_config)]

        # Warning: folder has no video files at all (maybe only .nfo or .jpg files)
        if not video_files:
            warnings.add(folder_rel, "No recognized video files found in folder")
            continue

        # Warning: video files found in a non-primary format (e.g. .mkv when primary is .mp4)
        for f in non_primary:
            _, ext = os.path.splitext(f.name)
            warnings.add(
                os.path.join(folder_rel, f.name),
                _format_primary_exts(media_config) + " video file — may need re-encoding",
                extension=ext.lower()
            )

        # Track seen editions within this folder to detect duplicates
        # Key = lowercased edition name, Value = filename of first file with that edition
        seen_editions = {}

        # Process each primary video file in the folder
        for vf in primary_files:
            stem, _ = os.path.splitext(vf.name)   # Strip the .mp4 extension
            parsed  = _parse_file_stem(stem)

            # Warning: file name doesn't follow Plex naming convention
            if not parsed:
                warnings.add(
                    os.path.join(folder_rel, vf.name),
                    "File name does not match Plex naming convention"
                )
                continue

            file_title, file_year, edition = parsed

            # Warning: empty edition tag found — {edition-} with nothing after the dash
            if edition == "":
                warnings.add(
                    os.path.join(folder_rel, vf.name),
                    "Empty edition tag found — {edition-} has no value after the dash"
                )
                edition = None   # Treat as no edition so the file still gets catalogued

            # Warning: suspicious year (before cinema existed, or in the future)
            if file_year < EARLIEST_FILM_YEAR or file_year > CURRENT_YEAR:
                warnings.add(
                    os.path.join(folder_rel, vf.name),
                    f"Suspicious year ({file_year}) — expected between {EARLIEST_FILM_YEAR} and {CURRENT_YEAR}"
                )

            if not parsed_folder:
                # Warning: the parent folder name doesn't follow Plex convention
                warnings.add(folder_rel, "Folder name does not match Plex naming convention")
            else:
                folder_title, folder_year = parsed_folder

                # Warning: file title doesn't match the folder it lives in
                if file_title.lower() != folder_title.lower():
                    warnings.add(
                        os.path.join(folder_rel, vf.name),
                        f"File title '{file_title}' does not match folder title '{folder_title}'"
                    )

                # Warning: file year doesn't match the folder year (checked independently of title)
                if file_year != folder_year:
                    warnings.add(
                        os.path.join(folder_rel, vf.name),
                        f"File year ({file_year}) does not match folder year ({folder_year}) in '{movie_folder.name}'"
                    )

            # Warning: two files in the same folder claim the same edition
            edition_key = (edition or "").lower()
            if edition_key in seen_editions:
                warnings.add(
                    os.path.join(folder_rel, vf.name),
                    f"Duplicate edition — another file in this folder already claims "
                    f"{'no edition' if not edition else repr(edition)}: '{seen_editions[edition_key]}'"
                )
            else:
                seen_editions[edition_key] = vf.name   # Record this edition as seen

            # Add the movie to records, or update an existing entry
            key = _make_key(file_title, file_year, edition)
            if key not in records:
                records[key] = {
                    "title":     file_title,
                    "year":      file_year,
                    "edition":   edition,
                    "qualities": set()   # set() automatically handles duplicates
                }
            records[key]["qualities"].add(tag)   # e.g. add "UHD" to this movie's quality set

    return records


# ─────────────────────────────────────────────
# Quality order
# ─────────────────────────────────────────────

# Populated at startup by init_quality_order() so qualities always appear
# in the same order as defined in config.json (e.g. UHD before HD before SD)
_QUALITY_ORDER = []

def init_quality_order(media_folders):
    """Called by scan.py at startup to set the canonical quality sort order."""
    global _QUALITY_ORDER
    _QUALITY_ORDER = [qf["tag"] for qf in media_folders]

def _quality_order(qualities_set):
    """Convert a set of quality tags to a sorted list matching config order."""
    return [q for q in _QUALITY_ORDER if q in qualities_set]


# ─────────────────────────────────────────────
# Serializer (called by core/scanner.py)
# ─────────────────────────────────────────────

def serialize(records):
    """
    Convert the internal records dict into a sorted list of dicts for JSON output.
    JSON doesn't support Python sets, so qualities is converted to a sorted list here.
    Records are sorted alphabetically by title, then year, then edition.
    """
    return sorted(
        [
            {
                "title":     r["title"],
                "year":      r["year"],
                "edition":   r["edition"],
                "qualities": _quality_order(r["qualities"])
            }
            for r in records.values()
        ],
        key=lambda x: (x["title"].lower(), x["year"], x["edition"] or "")
    )


# ─────────────────────────────────────────────
# DB writer (called by core/scanner.py)
# ─────────────────────────────────────────────

def write_db(conn, records):
    """
    Write movie records to the SQLite database.

    Uses an upsert pattern (INSERT ... ON CONFLICT DO UPDATE) so re-running
    the scanner updates existing rows rather than erroring on duplicates.
    The UNIQUE constraint on (title, year, edition) is what triggers the conflict.
    """
    cur = conn.cursor()

    # Create the movies table if it doesn't already exist
    cur.execute("""
        CREATE TABLE IF NOT EXISTS movies (
            id        INTEGER PRIMARY KEY AUTOINCREMENT,
            title     TEXT    NOT NULL,
            year      INTEGER NOT NULL,
            edition   TEXT,
            qualities TEXT    NOT NULL,
            UNIQUE(title, year, edition)
        )
    """)

    for r in records.values():
        # SQLite doesn't have a list type, so qualities is stored as a JSON string
        # e.g. '["UHD", "HD"]' — parse it with json.loads() when reading
        qualities_json = json.dumps(_quality_order(r["qualities"]))
        cur.execute("""
            INSERT INTO movies (title, year, edition, qualities)
            VALUES (?, ?, ?, ?)
            ON CONFLICT(title, year, edition) DO UPDATE SET
                qualities = excluded.qualities
        """, (r["title"], r["year"], r["edition"], qualities_json))

File: src/media/movies.ts
/**
 * media/movies.ts
 * ---------------
 * Movie-specific parsing, serialization, and DB logic for MOASYS-Vault.
 *
 * Expected Plex folder structure (default rules):
 *   <media_folder>/
 *     <Movie Title (YEAR)>/
 *       <Movie Title (YEAR)>.mp4
 *       <Movie Title (YEAR)> {edition-Edition Name}.mp4
 *
 * The patterns and constants that define those conventions live in
 * src/core/rules/movies.ts (with optional YAML overrides in rules/movies.yaml).
 * This file no longer hardcodes regexes or year ranges — it consumes them
 * through the factory below.
 */

import fs from 'fs'
import path from 'path'

import {
  MoviesConfig,
  MediaFolder,
  MovieRecord,
  MovieOutput,
  WarningCollector,
  MediaModule,
} from '../core/types'
import { hasExtension, isPrimary, formatPrimaryExts, findUnexpectedEntries } from '../core/files'
import { MoviesRules } from '../core/rules/movies'
import { compilePattern } from '../core/rules/helpers'

// ─────────────────────────────────────────────
// Helpers
// ─────────────────────────────────────────────

/**
 * Parse a Plex movie folder name using the configured folder pattern.
 * Returns { title, year } or null if the name doesn't match the pattern.
 *
 * The pattern is expected to use named capture groups `title` and `year`.
 */
function parseFolder(name: string, folderRegex: RegExp): { title: string; year: number } | null {
  const m = folderRegex.exec(name)
  if (!m?.groups) return null
  const { title, year } = m.groups
  if (title === undefined || year === undefined) return null
  return { title: title.trim(), year: parseInt(year, 10) }
}

/**
 * Parse a Plex movie file stem using the configured file pattern.
 * Returns { title, year, edition } or null.
 *
 * edition is null (no tag), "" (empty tag like `{edition-}`), or a string.
 * The pattern is expected to use named capture groups `title`, `year`, and
 * optionally `edition`.
 */
function parseFileStem(
  stem: string,
  fileRegex: RegExp
): { title: string; year: number; edition: string | null } | null {
  const m = fileRegex.exec(stem)
  if (!m?.groups) return null
  const { title, year, edition: editionRaw } = m.groups
  if (title === undefined || year === undefined) return null

  let edition: string | null
  if (editionRaw === undefined) {
    edition = null // No {edition-...} tag at all
  } else if (editionRaw.trim() === '') {
    edition = '' // Empty tag: {edition-}
  } else {
    edition = editionRaw.trim()
  }
  return { title: title.trim(), year: parseInt(year, 10), edition }
}

/**
 * Build a unique Map key for a movie record.
 * Lowercased so "The Crow" and "the crow" are treated as the same title.
 */
function makeKey(title: string, year: number, edition: string | null): string {
  return `${title.toLowerCase()}|${year}|${(edition ?? '').toLowerCase()}`
}

/** Return true if the qualities set matches one of the acceptable combos */
function isAcceptableCombo(qualities: Set<string>, combos: readonly string[][]): boolean {
  return combos.some(combo => combo.length === qualities.size && combo.every(q => qualities.has(q)))
}

/** Build the Plex-style movie name used as the warning path */
function movieDisplayName(record: MovieRecord): string {
  const base = `${record.title} (${record.year})`
  return record.edition ? `${base} {edition-${record.edition}}` : base
}

// ─────────────────────────────────────────────
// Factory
// ─────────────────────────────────────────────

/**
 * Create a movies media module bound to the provided rules.
 * Called once at startup by scan.ts after rules are loaded and validated.
 *
 * Rules are captured in closure rather than read from module-level state,
 * so the module is reusable and testable without globals (apart from the
 * qualityOrder slot already established by initTagOrder).
 */
export function createMoviesModule(
  rules: MoviesRules
): MediaModule<MovieRecord, MovieOutput, MoviesConfig> {
  // Compile pattern entries to RegExp objects once at boot. The rules schema
  // already validated that pattern and flags are well-formed, so no try/catch.
  const folderRegex = compilePattern(rules.patterns.folder)
  const fileRegex = compilePattern(rules.patterns.file)

  // year_range.max is resolved to a plain number by the loader, even if the
  // YAML wrote `max: current`. So this assertion is safe.
  const yearMin = rules.year_range.min
  const yearMax = rules.year_range.max as number

  let qualityOrder: string[] = []

  return {
    /** Set the canonical quality sort order from config */
    initTagOrder(mediaFolders: MediaFolder[]): void {
      qualityOrder = mediaFolders.map(mf => mf.tag)
    },

    /**
     * Walk one media folder (e.g. UHD/) and return a Map of movie records.
     * Called once per media_folder entry by core/scanner.ts.
     */
    scanMediaFolder(
      folderPath: string,
      folderName: string,
      tag: string,
      config: MoviesConfig,
      warnings: WarningCollector
    ): Map<string, MovieRecord> {
      const records = new Map<string, MovieRecord>()

      const rootEntries = fs.readdirSync(folderPath, { withFileTypes: true })

      // Loose video files at media folder level — silently dropped without
      // this check. Plex expects each movie inside a Title (YEAR)/ folder.
      if (rules.checks.warn_loose_files) {
        const looseRoot = rootEntries.filter(
          e => e.isFile() && hasExtension(e.name, config.video_extensions)
        )
        if (looseRoot.length > 0) {
          warnings.add(
            folderName,
            `${looseRoot.length} loose video file(s) in media folder root — Plex expects each movie inside a 'Movie Title (YEAR)' folder. ` +
              `Move each file into its own correctly-named folder.`
          )
        }
      }

      // Unexpected non-media, non-sidecar files at media folder root.
      if (rules.checks.warn_unexpected_entries) {
        const unexpected = findUnexpectedEntries(
          rootEntries,
          config.video_extensions,
          rules.sidecar_extensions
        )
        if (unexpected.length > 0) {
          const names = unexpected.map(e => `'${e.name}'`).join(', ')
          warnings.add(
            folderName,
            `Unexpected file(s) in media folder root: ${names}. ` +
              `Expected only Movie Title (YEAR)/ subfolders plus Plex sidecars (${rules.sidecar_extensions.join(', ')}).`
          )
        }
      }

      for (const entry of rootEntries) {
        if (!entry.isDirectory()) continue

        const movieFolderPath = path.join(folderPath, entry.name)
        const folderRel = path.join(folderName, entry.name)
        const parsedFolder = parseFolder(entry.name, folderRegex)

        let allFiles: fs.Dirent[]
        try {
          allFiles = fs.readdirSync(movieFolderPath, { withFileTypes: true })
        } catch {
          warnings.add(folderRel, 'Permission denied reading folder')
          continue
        }

        // Subfolders inside a movie folder are silently ignored — any video
        // files in them would be dropped from the catalog.
        if (rules.checks.warn_extra_subfolders) {
          const subfolders = allFiles.filter(e => e.isDirectory())
          if (subfolders.length > 0) {
            const names = subfolders.map(s => `'${s.name}'`).join(', ')
            warnings.add(
              folderRel,
              `Unexpected subfolder(s) in movie folder: ${names}. ` +
                `Plex expects all video files directly inside the Movie Title (YEAR)/ folder. ` +
                `Files inside these subfolders are not scanned.`
            )
          }
        }

        // Non-media, non-sidecar files inside a movie folder (e.g. .zip, .txt).
        if (rules.checks.warn_unexpected_entries) {
          const unexpected = findUnexpectedEntries(
            allFiles,
            config.video_extensions,
            rules.sidecar_extensions
          )
          if (unexpected.length > 0) {
            const names = unexpected.map(e => `'${e.name}'`).join(', ')
            warnings.add(
              folderRel,
              `Unexpected file(s) in movie folder: ${names}. ` +
                `Expected only video files plus Plex sidecars (${rules.sidecar_extensions.join(', ')}).`
            )
          }
        }

        const videoFiles = allFiles.filter(
          f => f.isFile() && hasExtension(f.name, config.video_extensions)
        )
        const nonPrimary = videoFiles.filter(f => !isPrimary(f.name, config))
        const primaryFiles = videoFiles.filter(f => isPrimary(f.name, config))

        if (videoFiles.length === 0) {
          if (rules.checks.warn_no_videos) {
            warnings.add(folderRel, 'No recognized video files found in folder')
          }
          continue
        }

        if (rules.checks.warn_non_primary) {
          for (const f of nonPrimary) {
            const ext = path.extname(f.name).toLowerCase()
            warnings.add(
              path.join(folderRel, f.name),
              `${formatPrimaryExts(config)} video file — may need re-encoding`,
              ext
            )
          }
        }

        // Track seen editions in this folder to detect duplicates
        const seenEditions = new Map<string, string>()

        for (const f of primaryFiles) {
          const stem = path.basename(f.name, path.extname(f.name))
          const parsed = parseFileStem(stem, fileRegex)

          if (!parsed) {
            if (rules.checks.warn_bad_file_name) {
              warnings.add(
                path.join(folderRel, f.name),
                'File name does not match Plex naming convention'
              )
            }
            continue
          }

          const { title: fileTitle, year: fileYear } = parsed
          let { edition } = parsed

          if (edition === '') {
            if (rules.checks.warn_empty_edition) {
              warnings.add(
                path.join(folderRel, f.name),
                'Empty edition tag found — {edition-} has no value after the dash'
              )
            }
            edition = null // Treat as no edition so file still gets catalogued
          }

          if (fileYear < yearMin || fileYear > yearMax) {
            if (rules.checks.warn_suspicious_year) {
              warnings.add(
                path.join(folderRel, f.name),
                `Suspicious year (${fileYear}) — expected between ${yearMin} and ${yearMax}`
              )
            }
          }

          if (!parsedFolder) {
            if (rules.checks.warn_bad_folder_name) {
              warnings.add(folderRel, 'Folder name does not match Plex naming convention')
            }
          } else {
            const { title: folderTitle, year: folderYear } = parsedFolder

            if (fileTitle.toLowerCase() !== folderTitle.toLowerCase()) {
              if (rules.checks.warn_title_mismatch) {
                warnings.add(
                  path.join(folderRel, f.name),
                  `File title '${fileTitle}' does not match folder title '${folderTitle}'`
                )
              }
            }

            if (fileYear !== folderYear) {
              if (rules.checks.warn_year_mismatch) {
                warnings.add(
                  path.join(folderRel, f.name),
                  `File year (${fileYear}) does not match folder year (${folderYear}) in '${entry.name}'`
                )
              }
            }
          }

          // Duplicate-edition check: two files in same folder claiming same edition
          const editionKey = (edition ?? '').toLowerCase()
          const existing = seenEditions.get(editionKey)
          if (existing !== undefined) {
            if (rules.checks.warn_duplicate_edition) {
              warnings.add(
                path.join(folderRel, f.name),
                `Duplicate edition — another file already claims ${!edition ? 'no edition' : `'${edition}'`}: '${existing}'`
              )
            }
          } else {
            seenEditions.set(editionKey, f.name)
          }

          const key = makeKey(fileTitle, fileYear, edition)
          if (!records.has(key)) {
            records.set(key, {
              title: fileTitle,
              year: fileYear,
              edition,
              qualities: new Set(),
            })
          }
          records.get(key)!.qualities.add(tag)
        }
      }

      return records
    },

    merge(existing: Map<string, MovieRecord>, incoming: Map<string, MovieRecord>): void {
      for (const [key, record] of incoming) {
        if (existing.has(key)) {
          for (const q of record.qualities) existing.get(key)!.qualities.add(q)
        } else {
          existing.set(key, record)
        }
      }
    },

    serialize(records: Map<string, MovieRecord>): MovieOutput[] {
      const orderQualities = (qs: Set<string>) => qualityOrder.filter(q => qs.has(q))

      return [...records.values()]
        .map(r => ({
          title: r.title,
          year: r.year,
          edition: r.edition,
          qualities: orderQualities(r.qualities),
        }))
        .sort((a, b) => {
          const t = a.title.toLowerCase().localeCompare(b.title.toLowerCase())
          if (t !== 0) return t
          if (a.year !== b.year) return a.year - b.year
          return (a.edition ?? '').localeCompare(b.edition ?? '')
        })
    },

    /**
     * Post-merge check: emit a warning for each movie that exists in more than
     * one quality folder, unless its quality set is in acceptable_quality_combos.
     */
    postScan(records: Map<string, MovieRecord>, warnings: WarningCollector): void {
      if (!rules.checks.warn_multi_quality) return

      for (const record of records.values()) {
        if (record.qualities.size <= 1) continue
        if (isAcceptableCombo(record.qualities, rules.acceptable_quality_combos)) continue

        const ordered = qualityOrder.filter(q => record.qualities.has(q))
        warnings.add(
          movieDisplayName(record),
          `Movie exists in multiple quality folders: ${ordered.join(', ')}`
        )
      }
    },
  }
}
